text/0000-implied-derive.md

```diff
@@ -0,0 +1,425 @@
+
+## Changes
+
+The transitive super-trait closures of all derivable standard library traits are:
+
+```
+Copy => {Clone}
+Clone => {}
+Ord => {Eq, PartialEq, PartialOrd}
+PartialOrd => {PartialEq}
+Eq => {PartialEq}
+PartialEq => {}
+Hash => {}
+Default => {}
+Debug => {}
+```
+
+For each of these traits `T` on the LHS, and all future derivable traits,
+`#[derive(T)]` will now imply `#[derive(C)]` for all `C` in the transitive
+super-trait closure of `T`.
+
+### Union, not concatenation
+
+When the compiler is dealing with the union of all explicitly derived and
+implicitly derived traits, the union of those traits will be derived instead of
+the concatenation. This is particularly important for two reasons:
+
+1. To minimize the breakage in transitioning to edition 2018.
+   If you in the future do `#[derive(Copy, Clone)]`,
+   then the deriving system will see this as `#[derive(Copy, Clone)]`
+   and not `#[derive(Copy, Clone, Clone)]`.
+
+2. `Eq` and `PartialOrd` both imply `PartialEq`, and so if the concatenation is
+   used rather than the union, then the compiler will derive `PartialEq` twice
+   which will result in overlapping impls.
+
+### `#[structural_match]`
+
+[RFC 1445]: https://github.com/rust-lang/rfcs/pull/1445
+
+This RFC supercedes [RFC 1445] in that `#[derive(Eq)]`, where transitively
+applied to the transitive closure of all field types of a type, implies
+`#[structural_match]`. In other words, the requirement of transitively having
+`#[derive(PartialEq, Eq)]` is reduced to `#[derive(Eq)]`.
+
+### Interaction with `#[derive_no_bound]` et al.
+
+[RFC 2353]: https://github.com/rust-lang/rfcs/pull/2353
+
+To ensure consistency, `#[derive_no_bound(Trait)]` as described in [RFC 2353]
+should also imply `#[derive_no_bound(Super)]` where `Super` is the
+transitive super-trait closure of `Trait`. Conversely,
+`#[derive_field_bound(Trait)]` will should imply `#[derive_field_bound(Super)]`.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+There are two main drawbacks of this proposal.
+
+## Breakage
+
+This RFC will break some code. Specifically, if a user has already manually
+given an `impl` for a trait in the transitive closure of super-traits for a
+particular trait, then an error will be raised. An example of this situation is:
+
+```rust
+#[derive(Copy)]
+struct Foo;
+
+impl Clone for Foo {
+    fn clone(&self) -> Self { Self {} }
+}
+```
+
+which, in the future, would be equivalent to:
+
+```rust
+#[derive(Copy, Clone)]
+struct Foo;
+
+impl Clone for Foo {
+    fn clone(&self) -> Self { Self {} }
+}
+```
+
+and thus resulting in the error:
+
+```rust
+error[E0119]: conflicting implementations of trait `std::clone::Clone` for type `Foo`:
+ --> src/main.rs:2:20
+  |
+2 |     #[derive(Copy, Clone)]
+  |                    ^^^^^ conflicting implementation for `Foo`
+...
+5 |     impl Clone for Foo {
+  |     ------------------ first implementation here
+```
+
+### Mitigating factors
+
+1. We can do this breakage as part of edition 2018.
```

Comment:
"can" != "should"

I really do not want to see "there's a new edition" be used as an excuse to make breaking changes that would not otherwise be considered. If we go down that path, IMO Rust has failed at its stability promises.

Reply:
> I really do not want to see "there's a new edition" be used as an excuse to make breaking changes that would not otherwise be considered.

I'm not sure what you mean by "would not otherwise be considered". This is a change we've wanted to make for a long time -- dating from https://github.com/rust-lang/rfcs/pull/1028 -- but we backed off last time because of stability concerns. Part of the point of an edition (not the whole point, but an important point) was to let us revisit questions like this, where the expected impact is low (and can of course be mitigated via lints and the usual transition story). Anyway, I don't like arguing in GH line comments, so I'll leave it at that, but perhaps below I'll try to make the 'positive case' for this change a bit stronger (i.e., why it's valuable from an ergonomics and learnability POV).

Reply:
But how does your comment from [back then](https://github.com/rust-lang/rfcs/pull/1028#issuecomment-91566317) not apply anymore?